### PR TITLE
Fix duplicate CSS in `dev` mode

### LINKED ISSUE #4102: 🐛 BUG: CSS is duplicated in dev server.
### What version of `astro` are you using?

1.0.0-rc.3

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

Mac OS X 12.5

#### Expected result:

When I access my website via the dev server, each stylesheet will be applied once.

#### Actual result:

Each stylesheet is applied twice — both component stylesheets, and global stylesheets imported via ESM — via duplicate `<style>` tags in the `<head>`:

Screen Shot 2022-07-30 at 12 13 29 PM

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/astro-dupe-css?file=src%2Flayouts%2Fhome.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Closes #4102
- During `dev`, Vite treats all `.css` files as JS modules that inject themselves into the page.
- This is great for HMR, but it leads to flashes of unstyled content during `astro dev`.
- To avoid this, Astro started injecting the static CSS while server-side rendering the file. This avoids any flashes of unstyled content, but when the CSS is injected for HMR, it leads to duplicate styles. 
- This PR adds a tiny MutationObserver to our `client/hmr` logic that removes any SSR'd styles as soon as the HMR styles are present on the client. No more duplicates!
- Also removed the `data-astro-injected` attribute because it's no longer needed!

## Testing

Tests using `data-astro-injected` updated. HMR already covered by existing tests.

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/runtime/client/hmr.ts [hunk 1]
```diff
@@ -1,21 +1,24 @@
...
-				const injectedStyle = document.querySelector(
-					`style[data-astro-injected="${file.acceptedPath}"]`
-				);
-				if (injectedStyle) {
-					injectedStyle.parentElement?.removeChild(injectedStyle);
+	// Vite injects `<style type="text/css">` for ESM imports of styles
+	// but Astro also SSRs with `<style>` blocks. This MutationObserver
+	// removes any duplicates as soon as they are hydrated client-side.
+	const injectedStyles = getInjectedStyles();
+	const mo = new MutationObserver((records) => {
```
Comment (FredKSchott) [inline packages/astro/src/runtime/client/hmr.ts:8]: overall not against this, but is there a simpler way that doesn't involve a mutation observer? In my head, I was thinking the HMR'd style-script could be responsible for removing it's duplicate `<style>` from the page as soon as it loads.

**Review by FredKSchott — APPROVED** [review]
LGTM. I'd love to see a simpler solution if you think one exists, but only if it's a light lift. Otherwise, this is more than enough heading into a great v1.0 release!

File: packages/astro/src/runtime/client/hmr.ts (hunk 1 above)
Comment (natemoo-re) [reply packages/astro/src/runtime/client/hmr.ts:8]: Unfortunately not without hacking how Vite handles styles out of the box. The existing approach was to only remove the original SSR'd style when we saw an HMR update come in, but the duplicate styles exist **until** an HMR happens, and in most cases it's reasonable that an HMR update will never happen. 

This approach also allows us to get rid of `data-astro-injected="src/styles/index.css"` all over our dev output. Overall this seemed like the most resilient and unobtrusive solution.

**Review by matthewp — APPROVED** [review]

**Comment by ChrisSki** [comment]
Any chance this made it's way back into dev mode? I'm experiencing duplicate css when using the following in an astro file:

```
<style lang="scss">
  @import "./BlogPost.scss";
</style>
```

**Comment by benmcgarvey** [comment]
I'm also seeing this when importing scss modules into React components. E.g.
```jsx
import styles from './navigation.module.scss';
```

They appear to be duplicated between inline and the original file

**Pull request merged**